pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/window/value/FirstValueWindowFunction.java

```diff
@@ -78,12 +86,78 @@ private List<Object> processRowsWindow(List<Object[]> rows) {
       lowerBound++;
       upperBound = Math.min(upperBound + 1, numRows - 1);
     }
+
+    return result;
+  }
+
+  private List<Object> processRowsWindowIgnoreNulls(List<Object[]> rows) {
+    int numRows = rows.size();
+    int lowerBound = _windowFrame.getLowerBound();
+    int upperBound = Math.min(_windowFrame.getUpperBound(), numRows - 1);
+
+    int indexOfFirstNonNullValue = -1;
+    // Find first non-null value in the first window
+    for (int i = Math.max(lowerBound, 0); i <= upperBound; i++) {
+      Object value = extractValueFromRow(rows.get(i));
+      if (value != null) {
+        indexOfFirstNonNullValue = i;
+        break;
+      }
```

Comment:
This code is also repeated when the lower bound is moved. I think it is worth it to move the code to its own function. Something like `findFirstNotNullInWindow(rows, lowerBound, upperBound)`. Not only makes the code easier to read but also makes the job easier for the jit trying to optimize the loop.